src/dc_apeerstate.rs

```diff
@@ -17,328 +18,307 @@ use crate::x::*;
-        (*peerstate).to_save |= 0x1i32;
+
+    if message_time > peerstate.last_seen_autocrypt {
+        peerstate.last_seen = message_time;
+        peerstate.last_seen_autocrypt = message_time;
+        peerstate.to_save |= 0x1i32;
         if (header.prefer_encrypt == EncryptPreference::Mutual
             || header.prefer_encrypt == EncryptPreference::NoPreference)
-            && header.prefer_encrypt.to_i32().unwrap() != (*peerstate).prefer_encrypt
+            && header.prefer_encrypt.to_i32().unwrap() != peerstate.prefer_encrypt
         {
-            if (*peerstate).prefer_encrypt == 1i32
+            if peerstate.prefer_encrypt == 1i32
                 && header.prefer_encrypt != EncryptPreference::Mutual
             {
-                (*peerstate).degrade_event |= 0x1i32
+                peerstate.degrade_event |= 0x1i32
             }
-            (*peerstate).prefer_encrypt = header.prefer_encrypt.to_i32().unwrap();
-            (*peerstate).to_save |= 0x2i32
-        }
-        if (*peerstate).public_key.is_null() {
-            (*peerstate).public_key = dc_key_new()
+            peerstate.prefer_encrypt = header.prefer_encrypt.to_i32().unwrap();
+            peerstate.to_save |= 0x2i32
         }
-        if 0 == dc_key_equals((*peerstate).public_key, (*header).public_key) {
-            dc_key_set_from_key((*peerstate).public_key, (*header).public_key);
+
+        if peerstate.public_key.as_ref() == Some(&header.public_key) {
+            peerstate.public_key = Some(header.public_key.clone());
             dc_apeerstate_recalc_fingerprint(peerstate);
-            (*peerstate).to_save |= 0x2i32
+            peerstate.to_save |= 0x2i32;
         }
-    };
+    }
 }
 
 pub unsafe fn dc_apeerstate_apply_gossip(
-    mut peerstate: *mut dc_apeerstate_t,
+    peerstate: &mut dc_apeerstate_t,
     gossip_header: &Aheader,
     message_time: time_t,
 ) {
-    if peerstate.is_null()
-        || (*peerstate).addr.is_null()
-        || (*(*gossip_header).public_key).binary.is_null()
-        || CStr::from_ptr((*peerstate).addr)
+    if peerstate.addr.is_null()
+        || CStr::from_ptr(peerstate.addr)
             .to_str()
             .unwrap()
             .to_lowercase()
             != gossip_header.addr.to_lowercase()
     {
         return;
     }
-    if message_time > (*peerstate).gossip_timestamp {
-        (*peerstate).gossip_timestamp = message_time;
-        (*peerstate).to_save |= 0x1i32;
-        if (*peerstate).gossip_key.is_null() {
-            (*peerstate).gossip_key = dc_key_new()
-        }
-        if 0 == dc_key_equals((*peerstate).gossip_key, (*gossip_header).public_key) {
-            dc_key_set_from_key((*peerstate).gossip_key, (*gossip_header).public_key);
+
+    if message_time > peerstate.gossip_timestamp {
+        peerstate.gossip_timestamp = message_time;
+        peerstate.to_save |= 0x1i32;
+        if peerstate.gossip_key.as_ref() == Some(&gossip_header.public_key) {
```

Comment:
Ditto. Not sure if it's a problem but similar logic as above.

Reply:
yep, definitely worng